lib/charms/cloudflare_configurator/v0/cloudflared_route.py

```diff
@@ -0,0 +1,132 @@
+                only existing cloudflared-route relation.
+        """
+        if not relation:
+            relation = self._charm.model.get_relation(relation_name=self._relation_name)
+        relation_data = relation.data[self._charm.app]
+        secret_id = relation_data.get(_TUNNEL_TOKEN_SECRET_ID_FIELD)
+        if not secret_id:
+            secret = self._charm.app.add_secret({_TUNNEL_TOKEN_SECRET_VALUE_FIELD: tunnel_token})
+            secret.grant(relation)
+            relation_data[_TUNNEL_TOKEN_SECRET_ID_FIELD] = secret.id
+        else:
+            secret = self._charm.model.get_secret(id=secret_id)
+            secret.set_content({_TUNNEL_TOKEN_SECRET_VALUE_FIELD: tunnel_token})
+
+    def unset_tunnel_token(self, relation: ops.Relation | None = None) -> None:
+        """Unset cloudflared tunnel-token in the integration.
+
+        Args:
+            relation: The relation to remote the tunnel-token from, if the relation is None, using
+                the only existing cloudflared-route relation.
+        """
+        if not relation:
+            relation = self._charm.model.get_relation(relation_name=self._relation_name)
+        data = relation.data[self._charm.app]
+        secret_id = data.get(_TUNNEL_TOKEN_SECRET_ID_FIELD)
+        if secret_id:
+            self._charm.model.get_secret(id=secret_id).remove_all_revisions()
+        data[_TUNNEL_TOKEN_SECRET_VALUE_FIELD] = ""
+
+    def _on_relation_broken(self, event: ops.RelationBrokenEvent):
+        self.unset_tunnel_token(event.relation)
+
+
+class CloudflaredRouteRequirer:
+    """cloudflared-route requirer."""
+
+    def __init__(
+        self, charm: ops.CharmBase, relation_name: str = DEFAULT_CLOUDFLARED_ROUTE_RELATION
+    ):
+        self._charm = charm
+        self._relation_name = relation_name
+
+    def get_tunnel_tokens(
+        self, from_relation: ops.Relation | list[ops.Relation] | None = None
+    ) -> list[str]:
+        """Get cloudflared tunnel-token from cloudflared-route integrations.
+
+        Args:
+            from_relation: relations to receive the tunnel-token from.
+
+        Returns:
+            A list of cloudflared tunnel-tokens.
+        """
+        tunnel_tokens = []
+        if from_relation:
+            relations = (
+                [from_relation] if isinstance(from_relation, ops.Relation) else from_relation
+            )
+        else:
+            relations = self._charm.model.relations[self._relation_name]
+        relations.sort(key=lambda r: r.id)
+        for relation in relations:
+            if not relation.app:
+                continue
+            relation_data = relation.data[relation.app]
+            secret_id = relation_data.get(_TUNNEL_TOKEN_SECRET_ID_FIELD)
+            if not secret_id:
+                continue
+            secret = self._charm.model.get_secret(id=secret_id)
+            tunnel_token = secret.get_content(refresh=True)[_TUNNEL_TOKEN_SECRET_VALUE_FIELD]
```

Comment:
Any error handling needed here to avoid the charm getting in `ErrorStatus`?

Reply:
This is intentional, as if an error is raised here, it means the charm library has a bug and should enter an error state.

Reply:
The other side of the integration could not be using the library and be sending bad data. Shouldn't that result in the charm being able to go to BlockedState saying that the interface is not being respected?

Reply:
In that case, the problem can't be handled in the `cloudflare-configurator` charm, as the issue needs to be fixed on the other side of the integration. I think in this case, an error state in `cloudflare-configurator` is also acceptable.